Add EmacsBehavior to CodeInput and add example

el-ethan opened the pull request:
Have CodeInput inherit EmacsBehavior and add "Key bindings" option to CodeInputTest so that users can try out the bindings.

akshayaurora commented on kivy/uix/codeinput.py:
```diff
@@ -85,6 +85,13 @@ class CodeInput(CodeNavigationBehavior, TextInput):
+
+    :attr:`active_key_bindings` is a :class:`~kivy.properties.StringProperty`
+    and defaults to ``'default'``
+    '''
```
..versionadded:: 1.9.2

akshayaurora commented on kivy/uix/behaviors/emacs.py:
```diff
@@ -44,6 +47,17 @@ class EmacsBehavior(object):
+    be enabled/disabled programmatically for widgets that inherit from
+    :class:`EmacsBehavior`. If the value is not ``'emacs'`` Emacs bindings will
+    be disabled.
+
```
..versionadded...

akshayaurora commented:
@el-ethan I think the vim side if this debate is absent... :)

Which behavior to have by default for code input... I feel this would be better served as a garden widget, where we can have EmacsCodeInput VIMCodeinput... instead of limiting to one type of behavior let's keep the core widget open for re-usability.

akshayaurora commented:
Oh, btw nice job!

dessant commented:
This repo is a great place for the behavior itself, but i agree with @akshayaurora that we should not enable it for codeinbut by default. Documenting the usage in the behavior docstring is enough.

el-ethan commented:
@akshayaurora, do you think the vim community would mind if VIMBehavior inherited from EmacsBehavior? Seriously though, I think it would be great if there was a VIMBehavior mixin as well - but better that it be written by a vim user :)

Also, even though CodeInput inherits from EmacsBehavior in this branch, by default the behavior doesn't change from that of kivy master. It will only change if you pass a different value for `active_key_bindings`. 

The reason why I initially added it to the CodeInput is that I was asked to add an example of EmacsBehavior to `examples/widgets/codeinput`, and this was what I came up with at the time. The `active_key_bindings` property allows the behavior of the CodeInput to be changed programmatically so that you don't have to decide what kind of behavior you want to use ahead of time. VIMBehavior could be added to CodeInput later, and used the same way. 

However, I am not so attached to this example, so if you want me to make some changes, or simply close this pull request, it is fine with me. Probably an example is not necessary anyway, since anyone who will use EmacsBehavior is likely already familiar with Emacs.

dessant commented on kivy/uix/codeinput.py:
```diff
@@ -52,7 +52,7 @@
 
 
-class CodeInput(CodeNavigationBehavior, TextInput):
+class CodeInput(CodeNavigationBehavior, EmacsBehavior, TextInput):
```
@el-ethan, we would like to merge your work. Could you remove `EmacsBehavior, ` from this line?

el-ethan commented:
I removed `EmacsBehavior` from `CodeInput` but I added made a few minor changes so that the example in `examples/widgets/codeinput` still works and can be further extended in the future with VIMBehavior, PicoBehavior ;),  etc. I can still remove the example altogether, if necessary.

dessant commented:
@el-ethan, thanks a lot! :+1:

dessant commented:
I've cleaned up a bit in https://github.com/kivy/kivy/pull/3894.

The pr limits EmacsBehavior (feature released in 1.9.1) usage to CodeInput only, before it could also be used with TextInput.

@team, should we move CodeInput.key_bindings to TextInput and preserve the original behavior?

tito commented on examples/widgets/codeinput.py:
```diff
@@ -3,12 +3,13 @@
 from kivy.core.text import LabelBase
 from pygments import lexers
-
+from pygame import font as fonts
```
This should be removed

dessant commented:
@tito, it was removed.

dessant commented:
ref https://github.com/kivy/kivy/pull/3898
